core/src/main/java/org/kiwix/kiwixmobile/core/utils/SharedPreferenceUtil.kt

```diff
@@ -0,0 +1,182 @@
+/*
+ * Kiwix Android
+ * Copyright (c) 2019 Kiwix <android.kiwix.org>
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package org.kiwix.kiwixmobile.core.utils
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.preference.PreferenceManager
+import androidx.appcompat.app.AppCompatDelegate
+import androidx.core.content.ContextCompat
+import io.reactivex.Flowable
+import io.reactivex.processors.PublishProcessor
+import org.kiwix.kiwixmobile.core.CoreApp.Companion.instance
+import org.kiwix.kiwixmobile.core.NightModeConfig
+import org.kiwix.kiwixmobile.core.NightModeConfig.Mode.Companion.from
+import java.io.File
+import java.util.HashSet
+import java.util.Locale
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Manager for the Default Shared Preferences of the application.
+ */
+@Singleton
+class SharedPreferenceUtil @Inject constructor(context: Context?) {
+  private val sharedPreferences: SharedPreferences =
+    PreferenceManager.getDefaultSharedPreferences(context)
+  val prefStorages = PublishProcessor.create<String>()
+  private val textZooms = PublishProcessor.create<Int>()
+  private val nightModes = PublishProcessor.create<NightModeConfig.Mode>()
+
+  val prefWifiOnly: Boolean
+    get() = sharedPreferences.getBoolean(PREF_WIFI_ONLY, true)
+
+  val prefHideToolbar: Boolean
+    get() = sharedPreferences.getBoolean(PREF_HIDE_TOOLBAR, true)
+
+  val prefIsFirstRun: Boolean
+    get() = sharedPreferences.getBoolean(PREF_IS_FIRST_RUN, true)
+
+  val prefFullScreen: Boolean
+    get() = sharedPreferences.getBoolean(PREF_FULLSCREEN, false)
+
+  val prefBackToTop: Boolean
+    get() = sharedPreferences.getBoolean(PREF_BACK_TO_TOP, false)
+
+  val prefNewTabBackground: Boolean
+    get() = sharedPreferences.getBoolean(PREF_NEW_TAB_BACKGROUND, false)
+
+  val prefExternalLinkPopup: Boolean
+    get() = sharedPreferences.getBoolean(PREF_EXTERNAL_LINK_POPUP, true)
+
+  val prefLanguage: String
+    get() = sharedPreferences.getString(PREF_LANG, Locale.ROOT.toString())
```

Comment:
* All these getters can be removed to oneliner assignment of vals. Is this change(removal of getters) needed? If yes or no, then how did you decide this?
* I went through the kt. docs that you latched onto my previous PR but couldn't comprehend anything case specific regarding usage of assignment to vals and vars, they just said to use `get()` when something is being computed, quite vague though.

Reply:
No.

```
val aTime= currentTimeInMillis()
print(aTime)
wait(500ms)
print(aTime)
```

if this program was run at time `0` the output is

```
0
0
```

If you need the code on the right hand side of the `=` to be run every time you access the variable then you want to use a getter.

```
val aTime get() = currentTimeMillis()
```
or
```
fun aTime() = currentTimeMillis()
```
 are basically the same thing and would produce on our example program

```
0
500
```

If you are trying to figure this stuff out I would recommend starting a scratch `ctrl+alt+shift+insert` and trying to play with some variables. The [playground](https://play.kotlinlang.org/) can also be a helpful learning tool and it has the latest milestone for kotlin 1.4 easily accessible in the settings

Reply:
So our sharedPrefs are slightly annoying because vars would typically suit them eg

```
var myPref
get()= sharePref.get(..etc
set()= sharePref.edit(..etc
```

but some of our preferences are only written by the system in our PrefsFragment hence these vals existing with no setter counterpart to their get.

Another option that comes up frequently for sharepreferences is using [delegated properties](https://kotlinlang.org/docs/reference/delegated-properties.html) for shareprefs. A bit more advanced and probably won't be universally applicable to all these preferences, google should provide many articles as it is quite popular since a talk at droidcon in '18

Reply:
A big thanks to you for such an exhaustive explaination. :guitar: